design-proposals/volume-claim-templates.md

```diff
@@ -0,0 +1,117 @@
+spec:
+  running: true
+  volumeClaimTemplates:
+  - metadata:
+      name: cirros-pvc
+    spec:
+    dataSourceRef:
+      apiGroup: cdi.kubevirt.io
+      kind: VolumeImportSource
+      name: my-import-source
+    resources:
+      requests:
+        storage: 1Gi
+  template:
+    spec:
+      domain:
+        devices: {}
+      volumes:
+      - persistentVolumeClaim:
+          name: cirros-pvc
+        name: cirros-disk
+```
+
+#### Open question:
+With datavolumestemplate one of the downsides was that we couldn't use generateName of the datavolume name.
+It will be cool if we could use it for the PVC name. Issue: we need to mention the PVC name in the vm.spec.template.volumes part when writing the VM spec. Any idea how to go past that? update the PVC name there after the PVC is created sounds like not the best solution. WDYT?
+
+### Virtctl create-pvc command
+
+Example for virtctl command uses:
+```bash
+$ virtctl create-pvc --generate-name=my-pvc- --size=1Gi
+```
+This command will create a PVC with a generated name with "my-pvc-" prefix. The PVC size will be 1Gi, it will have empty disk image and AccessModes and VolumeMode optimal for the defined default storage class.
+
+```bash
+$ virtctl create-pvc --name=my-pvc- --volume-import-source=my-import-source --size=1Gi --storage-class=rook-ceph-block
+```
+This command will create a PVC named "my-pvc" with AccessModes and VolumeMode optimal for rook-ceph-block. The PVC will be populated by the data source provided in VolumeImportSource CR called "my-import-source".
+
+```bash
+$ virtctl create-pvc --name=my-pvc --volume-clone-source=fedora
+```
+This will create a PVC my-pvc with DataSourceRef of VolumeCloneSource called my-source. CDI clone procedure doesn't require size, VolumeMode, AccessModes those will all be filled automatically.
+
+## Scalability
+I don't see any scalability issues.
+
+## Update/Rollback Compatibility
+Currently we will still support dataVolumeTemplates and dataVolumes but we do plan to eventually deprecate them.  
+We should update our docs and UI to use the new API and prevent from using or mentioning dataVolumes.
+
+## Functional Testing Approach
+Modify most tests that use dataVolumeTemplates to use volumeClaimTemplates.  
+Make sure this change integrates well with hotplugs, snapshot and restore and other storage related features.
+
+# Implementation Phases
+Currently Planned work:
+- Create a function in CDI API library which uses CDI existing code to generate an optimal PVC spec. The function will receive a partial PVC spec and will fill in (if missing) information such as VolumeMode and AccessModes according to [Storage Profile](https://github.com/kubevirt/containerized-data-importer/blob/main/doc/storageprofile.md)
+- Add Kubevirt new API of VolumeClaimTemplates which will be used to mention PVC source and basic information with end result of creating a PVC that will have the optimal properties and data.
+- New virtctl command for creating PVCs.
```

Comment:
I think the virtctl command should be regarded as optional.  It may make sense down the road but is not needed for the initial implementation.  We can also mention here that we considered a mutating webhook for PVCs to achieve some of this but due to some issues with that approach it is not being considered as a primary implementation goal.

Reply:
ok ill mention the virtctl command is optional for later.
Ill add the mutating webhook idea.